plugin/rename_file.py

```diff
@@ -0,0 +1,155 @@
+            return
+        sublime.set_timeout_async(lambda: self.run_async(old_path, new_path))
+
+    def run_async(self, old_path: str, new_path: str) -> None:
+        file_rename: FileRename = {
+            "newUri": filename_to_uri(new_path),
+            "oldUri": filename_to_uri(old_path)
+        }
+        Promise.all(list(self.create_will_rename_requests_async(file_rename))) \
+            .then(lambda responses: self.handle_rename_async(responses)) \
+            .then(lambda _: self.rename_path(old_path, new_path)) \
+            .then(lambda success: self.notify_did_rename(file_rename) if success else None)
+
+    def create_will_rename_requests_async(
+        self, file_rename: FileRename
+    ) -> Generator[Promise[tuple[WorkspaceEdit | None, weakref.ref[Session]]]]:
+        for session in self.sessions():
+            filters = session.get_capability('workspace.fileOperations.willRename.filters') or []
+            if match_file_operation_filters(filters, file_rename['oldUri']):
+                yield session.send_request_task(Request.willRenameFiles({'files': [file_rename]})) \
+                    .then(partial(lambda weak_session, response: (response, weak_session), weakref.ref(session)))
+
+    def is_case_change(self, path_a: str, path_b: str) -> bool:
+        return path_a.lower() == path_b.lower() and Path(path_a).stat().st_ino == Path(path_b).stat().st_ino
+
+    def handle_rename_async(self, responses: list[tuple[WorkspaceEdit | None, weakref.ref[Session]]]) -> Promise:
+        promises: list[Promise] = []
+        for response, weak_session in responses:
+            if (session := weak_session()) and response:
+                promises.append(session.apply_workspace_edit_async(response, is_refactoring=True))
+        return Promise.all(promises)
+
+    def rename_path(self, old: str, new: str) -> Promise[bool]:
+        old_path = Path(old)
+        new_path = Path(new)
+        restore_files: list[tuple[str, tuple[int, int], list[sublime.Region]]] = []
+        for view in self.window.views():
+            if (file_name := view.file_name()) and file_name.startswith(str(old_path)):
+                new_file_name = file_name.replace(str(old_path), str(new_path), 1)
+                restore_files.append((new_file_name, self.window.get_view_index(view), list(view.sel())))
+                if view.is_dirty():
+                    view.run_command('save', {'async': False})
+                view.close()  # LSP spec - send didClose for the old file
+        if (new_dir := new_path.parent) and not new_dir.exists():
+            new_dir.mkdir(parents=True)
+        try:
+            old_path.rename(new_path)
+        except Exception:
+            sublime.status_message("Unable to rename")
+            return Promise.resolve(False)
+
+        return Promise.all([
+            open_file_uri(self.window, file_name, group=group_index[0])
+            .then(partial(self.restore_view, selection, group_index))
```

Comment:
Just improving the style and naming (the tuple has items "group index" and "tab index" so calling the group "group_index" seems inaccurate):


```suggestion
            open_file_uri(self.window, file_name, group=group[0]).then(partial(self.restore_view, selection, group))
```